rust/otap-dataflow/crates/otap/src/experimental/gigla_exporter/gigla_client.rs

```diff
@@ -0,0 +1,357 @@
+
+use azure_core::credentials::TokenCredential;
+use azure_identity::{
+    DeveloperToolsCredential, DeveloperToolsCredentialOptions, ManagedIdentityCredential,
+    ManagedIdentityCredentialOptions, UserAssignedId,
+};
+use flate2::Compression;
+use flate2::write::GzEncoder;
+use reqwest::{
+    Client,
+    header::{AUTHORIZATION, CONTENT_ENCODING, CONTENT_TYPE, HeaderMap, HeaderValue},
+};
+use serde::Serialize;
+use std::io::Write;
+use std::sync::Arc;
+use std::time::Duration;
+
+use crate::experimental::gigla_exporter::config::{AuthMethod, Config};
+
+/// HTTP client for Azure Log Analytics Data Collection Rule (DCR) endpoint.
+///
+/// Handles authentication, compression, and HTTP communication with the Azure
+/// Monitor ingestion API.
+pub struct GigLaClient {
+    http_client: Client,
+    endpoint: String,
+    credential: Arc<dyn TokenCredential>,
+    scope: String,
+}
+
+impl GigLaClient {
+    /// Creates a new GigLA client instance from provided components.
+    ///
+    /// # Arguments
+    /// * `http_client` - The HTTP client to use for requests
+    /// * `endpoint` - The full endpoint URL for the GigLA ingestion API
+    /// * `credential` - The token credential for authentication
+    /// * `scope` - The OAuth scope for token acquisition
+    ///
+    /// # Returns
+    /// * `GigLaClient` - A configured client instance
+    pub fn from_parts(
+        http_client: Client,
+        endpoint: String,
+        credential: Arc<dyn TokenCredential>,
+        scope: String,
+    ) -> Self {
+        Self {
+            http_client,
+            endpoint,
+            credential,
+            scope,
+        }
+    }
+
+    /// Creates a new GigLA client instance from the configuration.
+    ///
+    /// # Arguments
+    /// * `config` - The GigLA exporter configuration
+    ///
+    /// # Returns
+    /// * `Ok(GigLaClient)` - A configured client instance
+    /// * `Err(String)` - Error message if initialization fails
+    pub fn new(config: &Config) -> Result<Self, String> {
+        let http_client = Client::builder()
+            .timeout(Duration::from_secs(30))
+            .build()
+            .map_err(|e| format!("Failed to create HTTP client: {e}"))?;
+
+        // Build the endpoint URL from config components
+        let endpoint = format!(
+            "{}/dataCollectionRules/{}/streams/{}?api-version=2021-11-01-preview",
+            config.api.dcr_endpoint, config.api.dcr, config.api.stream_name
+        );
+
+        // Create credential based on auth method in config
+        let credential: Arc<dyn TokenCredential> = match config.auth.method {
+            AuthMethod::ManagedIdentity => {
+                let mut options = ManagedIdentityCredentialOptions::default();
+
+                if let Some(client_id) = &config.auth.client_id {
+                    // User-assigned managed identity
+                    log::info!("Using user-assigned managed identity with client_id: {client_id}");
```

Comment:
We want to use `EffectHandler::info()` instead of `log::info!`.

Reply:
At the same time, that interface does not support message formatting. I am mildly unsatisfied with our logging options 😁 if we can't debug with message formatting. 

Moreover, I think we should prioritize doing something. I like the log::info! API and the usage here, this is idiomatic and ergonomic.

Reply:
Also, I just realized that `EffectHandler::info()` is meant to print stuff to console. `log::info!` is not for just that.

This particular usage of log statement doesn't seem to be doing structured logging. I think it's equivalent of creating a string using `format!` and passing it to `log::info!`. `log` crate does have support for structured logging though.

That said since we're using tokio runtime anyway, I think we should consider using tokio's `tracing` crate instead of `log`?

Reply:
I agree. Right now the overall situation for logging and tracing in our engine is not great. We need to improve it and pick a solution that adds the least possible overhead in our thread per core design. I don't have a firm recommendation yet. We need to run tests.